### PR TITLE
add keybindings framework 

### PR DESCRIPTION
**Pull request by kne42**
# Description

Add machinery for the adding, validation, and dispatch of keybindings on the viewer and its layers.

### Changes
- remove `on_key_press` and `on_key_release` methods from all layers
- on the viewer and its layers, store all initial keybinding information under the `default_keybindings` class variable, which is used to initialize the `keybindings` instance attribute
- add convenience methods for adding, removing, and rebinding keybindings

### key sequences
Key sequences are represented in the form `[modifier-]key`, e.g. `a`, `Control-c`, or `Control-Alt-Delete`. Valid modifiers are `Control`, `Alt`, `Shift`, and `Meta`, and can be represented as in Emacs with the shorthands `C`, `M`, `S`, and `s`, respectively.

Shift pressed with any letter will be consumed
to transform it to upper case, e.g. Shift + w => W.
Otherwise, the letter will be lower case.
Shift will also be consumed without transformation when pressed
with any other non-special key unless Control is also a modifier.

Special keys include Shift, Control, Alt, Meta, Up, Down, Left, Right,
PageUp, PageDown, Insert, Delete, Home, End, Escape, Backspace, F1,
F2, F3, F4, F5, F6, F7, F8, F9, F10, F11, F12, Space, Enter, and Tab

### function format
Functions take in only one argument: the parent that the function was attached to. This is the viewer or layer.

By default, all functions are assumed to work on key presses only, but can be denoted to work on release too by separating the function into two statements with the `yield` keyword:
```python
def hello_world(viewer):
    # on key press
    viewer.status = 'hello world!'

    yield

    # on key release
    viewer.status = 'goodbye world :('
```

### bind a key
```python
viewer.bind_key('n', load_next_image)

@viewer.bind_key('p')
def load_prev_image(viewer): ...
```

### unbind a key
```python
viewer.unbind_keys('n')
viewer.bind_key('p', None)
```

### rebind a key
```python
viewer.bind_key('Left', 'p')
```

### bind multiple keys
```python
viewer.bind_key(('n', load_next_image),
                ('p', load_prev_image))

keybindings = [('Right', 'n'),
               ('Left', 'p')]
viewer.bind_key(keybindings)

keybindings = {
    'Control-s': save,
    'Control-u', undo
}
viewer.bind_key(keybindings)

viewer.bind_key(a=None)
```

### unbind multiple keys
```python
viewer.unbind_keys('a', 'b', 'c')
viewer.unbind_keys(['a', 'b', 'c'])
viewer.unbind_keys()
```

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# References

# How has this been tested?

- [x] the test suite for my feature passes
- [x] all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
This is looking great - I love the syntax.

Can you update the `custom_key_bindings` example (maybe rename it to `custom_keybindings` too)

Also as we discussed earlier can you add a `reset_keybindings` method that will remove any custom keybindings and reset to the default ones

File: napari/components/_viewer/view/main.py
Comment (sofroniewn) [inline napari/components/_viewer/view/main.py]: can this have a more specific name like `key_generators` or something else

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [inline napari/layers/_shapes_layer/model.py:1595]: The "Escape" doesn't seem to be working for me - i.e. select the polygon tool, start drawing and then click escape - I just keep drawing the polygon and have to quit

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [inline napari/layers/_shapes_layer/model.py:1591]: `Control-c` and `Control-v` don't seem to be working

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [inline napari/layers/_shapes_layer/model.py:1580]: `Shift` doesn't seem to be working, i.e. select a shape and start resizing - the aspect ratio is no longer fixed

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [inline napari/layers/_shapes_layer/model.py:1594]: Backspace / delete isn't working either

**Review by sofroniewn — CHANGES_REQUESTED** [review]
Looking good, have requested some changes in my other comments

File: napari/components/_viewer/view/main.py
Comment (kne42) [reply napari/components/_viewer/view/main.py]: I've renamed it to `_key_release_generators`

**Comment by kne42** [comment]
All reported bugs have been fixed. I will note that there is now a bug with Qt key handlers themselves when it comes to the modifier keys `Control` (this is actually the Command key on Mac) and `Shift` being used together on OSX.

When you press `Shift + 2` you get `@` on QWERTY keyboards, which is what will show up in most cases with Qt e.g. `Alt + Shift + 2` -> `Alt-@`. However, attempting the same thing with `Control` yields a different result: `Control + Shift + 2` -> `Control-Shift-2`.

Please test again and tell me what other bugs you discover :)

**Comment by sofroniewn** [comment]
In `examples/custom_key_bindings.py` can you change `viewer.key_bindings` to `viewer.keybindings`

**Comment by sofroniewn** [comment]
I just gave this a spin, all looks good to go after you fix the example. Maybe @jni wants to give it a review before merging.

The bug that you mention with control and shift is strange, I'd ignore for now though, but well found!

**Review by sofroniewn — APPROVED** [review]

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py]: Public functions need docstrings. Come on, @kne42, you know this! =P

Comment (jni) [reply napari/util/keybindings.py]: (And it helps if private functions have docstrings also!)

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py]: All this logic is super complicated. I know you like metaprogramming but you should add clear and extensive comments that tell the rest of us mortals what exactly this is doing. :stuck_out_tongue_winking_eye: A docstring with a tiny example would be super useful.

File: napari/layers/_shapes_layer/model.py
Comment (jni) [inline napari/layers/_shapes_layer/model.py:1590]: I find it confusing that the functions are methods, but without `self` or `cls`... I also don't understand what to do if I want to set a keybinding to call a non-class function.

File: napari/layers/_shapes_layer/model.py
Comment (jni) [inline napari/layers/_shapes_layer/model.py:1472]: these all feel like they should be lambdas...?

File: napari/layers/_markers_layer/model.py
Comment (jni) [inline napari/layers/_markers_layer/model.py:517]: Super cool! =)

**Review by jni — CHANGES_REQUESTED** [review]
@kne42 at a minimum, I think this needs a lot more documentation, including docstrings, code comments, and an example of how to change the keybindings... Sorry that I haven't had time to test yet, will do soon!

**Comment by kne42** [comment]
This is just a WIP, comprehensive docs, examples, and tests as well as a cleaner implementation will all be added prior to merge. Please consult the TODO section on the PR for progress :)

File: napari/layers/_shapes_layer/model.py
Comment (kne42) [reply napari/layers/_shapes_layer/model.py:1590]: All functions take in their parent as an argument i.e. if you're adding a keybinding to the viewer you would get passed a viewer, if you're adding a keybinding to an image layer, you'll get that image layer.

Since on the methods, `self` refers back to the parent anyways, it just works out. If you have a preexisting function, you can just wrap it with a lambda: `lambda v: my_func()`.

File: napari/layers/_shapes_layer/model.py
Comment (kne42) [reply napari/layers/_shapes_layer/model.py:1472]: I was considering this but I think in the future we may want to have access to these functions in for example a YAML file (via some parsing):
```yaml
Shapes:
  Control-i: activate_vertex_insert_mode
```

**Comment by jni** [comment]
LOL Ok, I was just thrown off by @sofroniewn's approval. =) Ping me when ready for a re-review.

File: napari/layers/_shapes_layer/model.py
Comment (jni) [reply napari/layers/_shapes_layer/model.py:1472]: Side note: I'm not sure about separate sections in the yaml for each layer type, but in any case I would want it to include the full path to that class:

```yaml
napari.layers.Shapes:
    Ctrl-i: activate_vertex_insert_mode
```

calls `napari.layers.Shapes.activate_vertex_inside_mode`. (Note: the use of underscores should be consistent — which is a conundrum.)

I think we need to think long and hard about what callbacks accessible from the key config should get as input, *and this should always be consistent*. Is it the topmost layer? Is it the list of selected layers? Is it the Layers object itself? Is it the whole viewer?

I think it's ok for the bindings to change depending on the state of the viewer, but do we want the *inputs* to change depending on the state of the viewer? That could make it really difficult to keep everything straight as a plugin developer.

If we do define different callback inputs for different states, then we need to define what those states *can* be, and this should be a finite list.

I guess what you're going for here is that each layer can define its own shortcuts, and these shortcuts would be enabled in the case *and only in the case* where that is the only selected layer. In that case, the config file could look like:

```yaml
- base:
  - [base keybindings for all of napari]
- single-selection:
  - napari.layers.Shapes:
    - ctrl-i: path.to.my_function
- multi-selection:
  - crtl-up: napari.layers.shift_selected_layers_up
```

In that case, each top-level section could have different input arguments to the callbacks.

I really like the idea of full path specs on the functions, because then adding all kinds of crazy functionality could be added with a pip install and a .yaml file.

File: napari/components/_viewer/view/main.py
Comment (sofroniewn) [inline napari/components/_viewer/view/main.py:16]: I this `theme` stuff needed for the `keybindings` PR, or is it compensating for other refactoring, or can it be removed till later?

File: napari/layers/_labels_layer/model.py
Comment (sofroniewn) [inline napari/layers/_labels_layer/model.py:627]: I guess the reason for creating the dictionary instead of just using decorators here is so that we can reset to defaults later if we want, right?

File: napari/util/tests/test_keybindings.py
Comment (sofroniewn) [inline napari/util/tests/test_keybindings.py:1]: Will these tests now get automatically run as part of our CI?

File: napari/components/_viewer/view/main.py
Comment (kne42) [reply napari/components/_viewer/view/main.py:16]: That must've been my mistake when rebasing, will fix.

File: napari/layers/_labels_layer/model.py
Comment (kne42) [reply napari/layers/_labels_layer/model.py:627]: No, using decorators would also add it to the default keybindings anyways (as long as it is initialized as an empty dictionary). However, you cannot use the decorators in this situation because you don't have access to them – `bind_key` et al have not been defined in this scope; they are inherited from the base `Layer` class.

File: napari/util/tests/test_keybindings.py
Comment (kne42) [reply napari/util/tests/test_keybindings.py:1]: Yes, any file starting with `test_` will have all `test_*` methods run automatically by `pytest`.

File: napari/layers/_labels_layer/model.py
Comment (sofroniewn) [reply napari/layers/_labels_layer/model.py:627]: ah ok so the decorators are to be used by our users / plugin developers, but internally we will use the `default_keybindings` dictionary. that's fine

File: napari/util/tests/test_keybindings.py
Comment (sofroniewn) [reply napari/util/tests/test_keybindings.py:1]: oh great!

File: napari/layers/_labels_layer/model.py
Comment (kne42) [reply napari/layers/_labels_layer/model.py:627]: Not quite – I would say that we want to use the functional/decorator form wherever possible; it's just not possible to add these keybindings to a class within its own definition. I do know how to make it work but I'm not sure how useful it would really be.

**Comment by kne42** [comment]
Proper tests and docstrings have been added and this is ready for review. Helpful comments to include would be on the clarity of the documentation, intuitiveness of the API, and whether the API is too overloaded or not.

cc @gregjohnso

File: napari/components/_viewer/model.py
Comment (jni) [inline napari/components/_viewer/model.py:26]: Yeah, this sorta needs to go. =P

File: napari/components/_viewer/model.py
Comment (jni) [inline napari/components/_viewer/model.py:179]: What's the input here? please add a docstring.

File: napari/components/_viewer/model.py
Comment (jni) [inline napari/components/_viewer/model.py:190]: Can you add some comments to these definitions? Needless to say, this is not a standard way to define methods, so some comments explaining the rationale here would be great.

Some thoughts that may or may not be relevant:
- should `keybindings=` be renamed `instance_keybindings=`?
- could this be more cleanly implemented using multiple inheritance? (ie make a `AcceptsKeybindings` class?)

File: napari/components/_viewer/view/main.py
Comment (jni) [reply napari/components/_viewer/view/main.py:16]: This doesn't seem fixed yet. =P

Also, the dreaded quadruple-dot has reared its ugly head. Perhaps this is not the PR to fix it but just want to flag again that I think our hierarchy is too deep and complex. `components` seems redundant, actually... Isn't everything a component?

Comment (jni) [reply napari/components/_viewer/view/main.py:16]: (Note, please don't fix this in this PR, it'll be a mess! :joy:)

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:21]: I think I've said this before, but this is badass. =D

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:31]: :joy:

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:86]: Are these shortcuts standards somewhere?

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:127]:
```suggestion
    *modifiers, key = parsed
```

Python 3, baby!

```python
In [1]: *foo, bar = [1, 2, 3]
In [2]: foo
Out[2]: [1, 2]
In [3]: bar
Out[3]: 3
```

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:132]: a) some doctest examples would be super useful
b) as discussed in the call, let's rename sequence to combination

In some sense it's hard to distinguish between components and a key combination. My suggestion would be something like `normalize_key_combo`.

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:172]: This note should probably be in the module-level docstring

File: napari/util/keybindings.py
Comment (jni) [reply napari/util/keybindings.py:132]: LMAO I just saw that that's the next function... Maybe I should stop using "Add single comment" =P

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:194]: Can you write a bit more about the shift-consuming choice? Is that standard? I kind of dislike it because some people might read `W` as just the key `w`...

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:224]: Can you elaborate on why this function exists? Where is it used? (I know I can go and look for usage but in general *code should be local*, ie you should not need to have global context in your head to understand the piece of code you're reading.

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:238]: As mentioned in the discussion, here and elsewhere, replace `seq` with `combo`

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:281]: I don't love `_rebind`... Can we do without it?

I'm also not sure that `bind_key` should ever unbind, but I can see how `bind_key(dict, 'r', None)` might be an intuitive way to deal with this. BUT, not with `unbind_key` around! There should be one way to do this, I think.

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:383]: This is a bit of a crazy API... I'm wondering if `unbind_keys` should not coexist with `unbind_key` — just have one function that deals with both use cases. I'd prefer if we didn't need to return `funcs` either... I can see where it might be useful, but we should try to keep our API minimal to begin with, and grow it where needed, rather than start with an enormous API.

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:503]: Wait what is this? I don't see the import for it...

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:582]: This is super funky logic... Can this not be simplified?

**Review by jni — CHANGES_REQUESTED** [review]
@kne42 this still feels too complex for me. I don't have a good sense at all of how the framework works. The example is super great, but then there's all this complexity behind the scenes that I think will be very hard to maintain. At the very least it needs a lot more documentation about how it all fits together. But my instinct is that all this stuff can be achieved with significantly less code. I've left comments in the specific places I thought were confusing.

**Comment by jni** [comment]
Also, are we leaving the yaml file for a future PR?

File: napari/util/keybindings.py
Comment (sofroniewn) [reply napari/util/keybindings.py:194]: this is tricky, most of the menu bars that I see just use capital letters and then the up arrow if the shift key is to be pressed, but i think using `w` and `W` also makes sense

File: napari/util/keybindings.py
Comment (sofroniewn) [reply napari/util/keybindings.py:281]: @jni does that mean that if you try and bind to an already bound key the effect would be additive, not to replace the existing binding? I'm not sure what is most intuitive, but best practice would probably be to let the user know somehow that there is already something there

**Comment by sofroniewn** [comment]
@jni I think `yaml` can wait to a future PR.

I think having `bind_key` which allows us to use the decorators such as `@viewer.bind_key('a')` is really nice, and having `unbind_keys` exposed on the main object also makes sense to me.

I'd love to see the `example/custom_keybindings.py` get fleshed out a bit more, to include adding custom keybindings to different layer instances (i.e. `'q'` on one image layer does something different then `'q'` on another image layer), and adding to the `Markers` class, i.e. `'w'` will do the same thing on all markers layers, even ones that havn't been created yet. It would be nice to see an example of `unbind_keys` in there too to unbind a single key and multiple keys. Maybe also an example using some of the modifier keys like `Shift` or `Command`, and then maybe an example using some of the special keys like `SpaceBar`, `LeftArrow`, and `RightArrow`. For all of these cases you can just have the function do something like print the layer name / viewer title and the key that was pressed.

File: napari/components/_viewer/model.py
Comment (kne42) [reply napari/components/_viewer/model.py:190]: Ah yeah, I think I may go with your multiple inheritance idea.

File: napari/util/keybindings.py
Comment (kne42) [reply napari/util/keybindings.py:127]: WHAT

Comment (kne42) [reply napari/util/keybindings.py:127]: mind officially blown. python best lang

File: napari/util/keybindings.py
Comment (kne42) [reply napari/util/keybindings.py:86]: Emacs teehee

File: napari/util/keybindings.py
Comment (kne42) [reply napari/util/keybindings.py:172]: it is already. did i not make it clear enough?

Comment (jni) [reply napari/util/keybindings.py:172]: @kne42 I don't know if I'm blind but [here's](https://github.com/napari/napari/pull/251/files#diff-266e2fe0a33dbefd726fb2bf1e11055bR4) the module-level docstring and I don't see anything about Ctrl being macOS Cmd...

File: napari/util/keybindings.py
Comment (jni) [reply napari/util/keybindings.py:127]: Obviously I agree, but I would be remiss not to point out that this kind of pattern-matching was copied from functional languages such as Haskell. =)

File: napari/util/keybindings.py
Comment (jni) [reply napari/util/keybindings.py:194]: > most of the menu bars that I see just use capital letters and then the up arrow if the shift key is to be pressed

Exactly. I think expressing the shortcut with the capital is kinda non-standard and therefore bad...

Comment (sofroniewn) [reply napari/util/keybindings.py:194]: @freeman-lab want to weigh in here?

File: napari/util/keybindings.py
Comment (jni) [reply napari/util/keybindings.py:281]: I'm happy for `bind` to always replace, OR for it to be additive (with a warning) and having to do unbind/bind in sequence. I just think we should keep this API small. Maybe even having just `bind(..., None)` being `unbind`, then the API is really just one function.

File: napari/util/keybindings.py
Comment (kne42) [reply napari/util/keybindings.py:194]: this was more to play it in-line with how Qt handles other key combinations

for example, pressing `Shift+2` actually gets read in by Qt as `Shift+@`

Comment (jni) [reply napari/util/keybindings.py:194]: I think you mean `Shift-2` gets read-in as `@`?

At any rate, we should aim for our user-facing documentation to be as simple and intuitive as possible, and not have it bound to implementation details. In this case, I think writing "Shift-2" as the shortcut is way more intuitive than writing `@` as the shortcut.

Comment (kne42) [reply napari/util/keybindings.py:194]: no, `Shift-2` gets read in as `Shift-@`. our cleaner functions then remove the `Shift` because it is applied to a non-special key so it appears to the user as `@`

there is absolutely no way for us to know that `2` was ever pressed due to regional keyboard differences

which is why `Shift-2` as the shortcut would be invalid; also if we did want that we would have to create a table of all of the different key combos and their unicode and convert them back (which again would only work on US keyboards)

Comment (jni) [reply napari/util/keybindings.py:194]: Fascinating! Yes, `@` is AltGr + 2 on ISO keyboards. What a mess.

Comment (jni) [reply napari/util/keybindings.py:194]: @kne42 I think the "right" way to do this is to set/display the key combo we want, e.g. "shift-2", and then process the key based on the OS/system keyboard setting. This might be super hard but there might be pre-existing libraries to deal with all this stuff?

**Comment by jni** [comment]
@kne42 will you get a chance to have another crack at this sometime soon?

File: napari/util/keybindings.py
Comment (kne42) [reply napari/util/keybindings.py:194]: seems way too complex to me. i think it's best that we leave this alone for now

**Comment by kne42** [comment]
@jni i am currently working on a new version at https://github.com/kne42/napari/blob/better-keybindings/napari/util/keybindings.py

**Comment by kne42** [comment]
closed in favor of #389

**Pull request closed without merging**
